File: Minimum Multiplications to reach End - GFG/minimum-multiplications-to-reach-end.cpp
//{ Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function Template for C++

class Solution {
  public:
   int mod = 100000;
    int minimumMultiplications(vector<int>& arr, int start, int end) {
        // code here
        queue<pair<int,int>> q;
        q.push({0,start});
        vector<int> dist(1e5,1e9);
        //as starting distance is 0;
        dist[start] = 0;
        while(!q.empty()){
            auto it = q.front();
            int op = it.first;
            int from = it.second;
            if(from==end) return op;
            q.pop();
            for(int i = 0;i<arr.size();i++){
                int to = (from*arr[i])%mod;
                if(op+1<dist[to]){
                    dist[to] = op + 1;
                    q.push({op+1,to});
                }
            }
        }
        return dist[end]==1e9?-1:dist[end];
    }
};


//{ Driver Code Starts.

int main() {

    int t;
    cin >> t;
    while (t--) {
        int n;
        cin >> n;
        vector<int> arr(n);
        for (int i = 0; i < n; i++) {
            cin >> arr[i];
        }
        int start, end;
        cin >> start >> end;
        Solution obj;
        cout << obj.minimumMultiplications(arr, start, end) << endl;
    }
}

// } Driver Code Ends

File: Medium/Minimum Multiplications to reach End/minimum-multiplications-to-reach-end.java
//{ Driver Code Starts
// Initial Template for Java

import java.util.*;
import java.lang.*;
import java.io.*;

// Position this line where user code will be pasted.

class GFG {
    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        int T = sc.nextInt();
        while (T-- > 0) {
            int n = sc.nextInt();
            int[] a = new int[n];

            for (int i = 0; i < n; i++) {
                a[i] = sc.nextInt();
            }

            int start = sc.nextInt();
            int end = sc.nextInt();

            Solution ob = new Solution();
            int ans = ob.minimumMultiplications(a, start, end);
            System.out.println(ans);
        }
    }
}
// } Driver Code Ends

class Solution {
    int minimumMultiplications(int[] arr, int start, int end) {
        Queue<Integer> q = new LinkedList<>();
        q.add(start);
        int[] dist = new int[100000];
        Arrays.fill(dist, Integer.MAX_VALUE);
        dist[start] = 0;
        
        while (!q.isEmpty()) {
            int node = q.poll();
            
            for (int num : arr) {
                int nextNode = (int) (((long) num * node) % 100000);
                
                if (dist[nextNode] > dist[node] + 1) {
                    dist[nextNode] = dist[node] + 1;
                    q.add(nextNode);
                }
            }
        }
        
        return dist[end] == Integer.MAX_VALUE ? -1 : dist[end];
    }
}
